Source/CarthageKit/Project.swift

```diff
@@ -1036,7 +1007,24 @@ public final class Project { // swiftlint:disable:this type_body_length
 				}
 			}
 			.flatMap(.concat) { dependencies -> SignalProducer<(Dependency, PinnedVersion), CarthageError> in
-				return .init(dependencies)
+				return SignalProducer(dependencies)
+					.flatMap(.concurrent(limit: 4)) { dependency, version -> SignalProducer<(Dependency, PinnedVersion), CarthageError> in
+						switch dependency {
+						case .git, .gitHub:
+							guard options.useBinaries else {
+								return .init(value: (dependency, version))
+							}
+							return self.installBinaries(for: dependency, pinnedVersion: version, toolchain: options.toolchain)
+								.filterMap { installed -> (Dependency, PinnedVersion)? in
+									return installed ? nil : (dependency, version)
+							}
+						case let .binary(url):
+							return self.installBinariesForBinaryProject(url: url, pinnedVersion: version, projectName: dependency.name, toolchain: options.toolchain)
+								.then(SignalProducer<(Dependency, PinnedVersion), CarthageError>.empty)
+						}
+					}
+					.collect()
+					.flatten()
```

Comment:
`collect` and `flatten` cancel each other out; these can be removed.

Reply:
Well, I thought we should wait all download processes to complete before building as some dependencies might need them to build. Isn't it correct?

Reply:
That's not necessary.

The dependencies are order from leaves to root—i.e. the dependencies always come first. The next line after this is `flatMap(.concat)`. Because we've specified the `concat` strategy, dependencies are built one at a time, sequentially. So it's not important that all the downloads finish before we hit that.

But actually… we need to keep the original order of `dependencies`, which might be affected by the `concurrent(limit: 4)`. So we should do `.then(SignalProducer(dependencies))` here instead.

Reply:
[0] A: Pre-built binary is available
[1] B: Pre-built binary is available
[2] C: Pre-built binary is unavailable, has dependency to A and B

In this case, it starts build C as soon as A _or_ B's download has finished.
So I think we need to wait for all the download to be done.

Reply:
But we also need to retain the original order, which this doesn't do. The order matters when building from scratch because each project's dependencies need to already have been built.

Reply:
Addressed https://github.com/Carthage/Carthage/pull/2342/commits/fabac4ce9226bb4cfdd5509224692f6962b4a6cc